crates/testing/tests/unreliable_network.rs

```diff
@@ -0,0 +1,322 @@
+)]
+#[cfg_attr(async_executor_impl = "async-std", async_std::test)]
+#[instrument]
+async fn libp2p_network_async() {
+    async_compatibility_layer::logging::setup_logging();
+    async_compatibility_layer::logging::setup_backtrace();
+    let metadata = TestMetadata {
+        overall_safety_properties: OverallSafetyPropertiesDescription {
+            check_leaf: true,
+            ..Default::default()
+        },
+        completion_task_description: CompletionTaskDescription::TimeBasedCompletionTaskBuilder(
+            TimeBasedCompletionTaskDescription {
+                duration: Duration::new(240, 0),
+            },
+        ),
+        unreliable_network: Some(Box::new(AsynchronousNetwork {
+            keep_numerator: 8,
+            keep_denominator: 10,
+            delay_low_ms: 4,
+            delay_high_ms: 30,
+        })),
+        ..TestMetadata::default_multiple_rounds()
+    };
+
+    metadata
+        .gen_launcher::<TestTypes, Libp2pImpl>()
+        .launch()
+        .run_test()
+        .await
+}
+
+#[cfg(test)]
+#[cfg_attr(
+    async_executor_impl = "tokio",
+    tokio::test(flavor = "multi_thread", worker_threads = 2)
+)]
+#[cfg_attr(async_executor_impl = "async-std", async_std::test)]
+async fn test_memory_network_async() {
+    use hotshot_testing::{
+        completion_task::{CompletionTaskDescription, TimeBasedCompletionTaskDescription},
+        node_types::{MemoryImpl, TestTypes},
+        test_builder::TestMetadata,
+    };
+    use std::time::Duration;
+
+    async_compatibility_layer::logging::setup_logging();
+    async_compatibility_layer::logging::setup_backtrace();
+    let metadata = TestMetadata {
+        // allow more time to pass in CI
+        completion_task_description: CompletionTaskDescription::TimeBasedCompletionTaskBuilder(
+            TimeBasedCompletionTaskDescription {
+                duration: Duration::from_secs(240),
+            },
+        ),
+        unreliable_network: Some(Box::new(AsynchronousNetwork {
+            keep_numerator: 8,
+            keep_denominator: 10,
+            delay_low_ms: 4,
+            delay_high_ms: 30,
+        })),
+        ..TestMetadata::default()
+    };
+    metadata
+        .gen_launcher::<TestTypes, MemoryImpl>()
+        .launch()
+        .run_test()
+        .await;
+}
+
+#[cfg(test)]
+#[cfg_attr(
+    async_executor_impl = "tokio",
+    tokio::test(flavor = "multi_thread", worker_threads = 2)
+)]
+#[cfg_attr(async_executor_impl = "async-std", async_std::test)]
+async fn test_memory_network_partially_sync() {
+    use hotshot_testing::{
+        completion_task::{CompletionTaskDescription, TimeBasedCompletionTaskDescription},
+        node_types::{MemoryImpl, TestTypes},
+        test_builder::TestMetadata,
+    };
+    use std::time::Duration;
+
+    async_compatibility_layer::logging::setup_logging();
+    async_compatibility_layer::logging::setup_backtrace();
+    let metadata = TestMetadata {
+        // allow more time to pass in CI
```

Comment:
Is the following `timeout_ms: 30` the one determine how tolerant we are for CI tests on determine whether a message (or actually it means a view?) is timed out?

Reply:
`timeout_ms` denotes the maximum delay of a message. For example, if I send message M from A to B, `B` will receive it in time `t` where `t` is upper bounded with `timeout_ms + time for network to send message`.

Reply:
Thanks for the clear illustration! Yeah either renaming or adding comment could help.